llama-index-core/llama_index/core/workflow/context.py

```diff
@@ -1,20 +1,87 @@
+import asyncio
 from typing import Dict, Any, Optional, List, Type
+from types import TracebackType
 
 from .events import Event
 
 
 class Context:
+    """A global object representing a context for a given workflow run.
+
+    The Context object can be used to store data that needs to be shared across iterations during a workflow execution.
+    Steps can use data stored in a Context object to keep a state across multiple executions within a workflow run.
+    Any Context instance offers two type of data storage: `ctx.globals`, that's shared among all the steps within a
+    workflow, and `ctx.locals`, that's private to a single step.
+
+    Note that using `ctx.globals` is not coroutine-safe if you `await` for something in-between accessing
+    the global state. In that case, make sure to lock the Context object before accessing `ctx.globals`. You can also
+    use the Context object as an async context manager:
+
+        with ctx:
+            ctx.globals["foo"] = "bar"
+            await some_async_call()
+            ctx.globals["foo"] = "baz"
+    """
+
     def __init__(self, parent: Optional["Context"] = None) -> None:
-        # Global state
+        # Global data storage
         if parent:
-            self.data = parent.data
+            self._globals = parent.data
         else:
-            self.data: Dict[str, Any] = {}
+            self._globals: Dict[str, Any] = {}
+            self._lock = asyncio.Lock()
 
-        # Step-specific instance
-        self.parent = parent
+        # Local data storage
+        self._locals: Dict[str, Any] = {}
+
+        # Local configuration
+        self._parent: Optional[Context] = parent
         self._events_buffer: Dict[Type[Event], Event] = {}
 
+    @property
+    def parent(self) -> Optional["Context"]:
```

Comment:
We might not want to allow parent context access

If we think of nesting workflows, coupling them too much will make testing harder for the developer no?

instead, I would encourage the parent workflow passing in whatever is needed to run the child workflow 

(I might be misunderstanding the intention of this too though)

Reply:
This is just to access the lock, I can make this property `def lock(self)` and avoid exposing the parent context 👍

Reply:
yea that might be a bit more straightforward 👀